### PR TITLE
feat: add new metadata to Datadog traces (attempt 2)

### PR DESCRIPTION
**Pull request by macaptain**
## Overview

Re-implementation of https://github.com/omgnetwork/elixir-omg/pull/1763

Add service, version and error metadata to traces in Datadog APM traces.

What you see in the traces:
* The service name (default 'web' for Phoenix applications) is the name of the OMG application.
* Traces are tagged with application version number
* For API responses which error, the trace is flagged red as an error, the error code is marked on the trace and the error message is found in the tags.

image

## Changes

- Move service name and version getters from the Response module to the Application module.
- Add a function for customizing trace metadata for Spandex which fetches details from conn.assigns and the service / version getters.
- Add two tests for the trace metadata function.

## Testing

```
DD_API_KEY=<some-api-key> docker-compose -f docker-compose-watcher.yml -f docker-compose.dev.yml up
make docker-update-watcher
make docker-update-watcher_info
```

Make API calls to the Watcher endpoints, and check the local-development environment APM Services in Datadog. Look for a service called "Watcher" and check the traces look as expected in the screenshot above.

### CODE REVIEW

File: apps/omg_watcher_rpc/test/omg_watcher_rpc/tracer_test.exs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  setup do
+    original_mode = Application.get_env(:omg_watcher_rpc, :api_mode)
+    _ = on_exit(fn -> Application.put_env(:omg_watcher_rpc, :api_mode, original_mode) end)
+
+    :ok
+  end
+
+  test "api responses without errors get traced with metadata" do
+    :ok = Application.put_env(@app, :api_mode, :watcher)
+    version = OMGApplication.version()
```
Comment (macaptain) [inline apps/omg_watcher_rpc/test/omg_watcher_rpc/tracer_test.exs:32]: Let me know if there's a better way to mock the call of `version()`.

File: apps/omg_watcher_rpc/lib/tracer.ex [hunk 2]
```diff
@@ -18,4 +18,26 @@ defmodule OMG.WatcherRPC.Tracer do
...
+
+    conn
+    |> SpandexPhoenix.default_metadata()
+    |> Keyword.put(:service, service_name)
+    |> Keyword.put(:error, [{:error, true}])
+    |> Keyword.put(:tags, [{:version, version}, {:"error.type", error_type}, {:"error.msg", error_msg}])
+  end
+
+  def add_trace_metadata(conn) do
+    service_name = Application.service_name()
```
Comment (InoMurko) [inline apps/omg_watcher_rpc/lib/tracer.ex:35]:
```suggestion
    service_name = OMG.WatcherRPC.Configuration.service_name()
    service_name = OMG.WatcherRPC.Configuration.version()
```
And if the configuration module doesn't exist, you can create one

Comment (macaptain) [reply apps/omg_watcher_rpc/lib/tracer.ex]: I've refactored these functions into a new Configuration module.

File: apps/omg_watcher_rpc/lib/application.ex [hunk 3]
```diff
@@ -17,6 +17,8 @@ defmodule OMG.WatcherRPC.Application do
   use Application
   use OMG.Utils.LoggerExt
 
+  @app :omg_watcher_rpc
```
Comment (InoMurko) [inline apps/omg_watcher_rpc/lib/application.ex:20]: I think this can be removed

**Review by InoMurko — APPROVED** [review]

File: apps/omg_watcher_rpc/test/omg_watcher_rpc/tracer_test.exs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.WatcherRPC.TracerTest do
+  @app :omg_watcher_rpc
+
+  use ExUnit.Case, async: true
```
Comment (InoMurko) [inline apps/omg_watcher_rpc/test/omg_watcher_rpc/tracer_test.exs:18]: your test is fiddling with global state `Application.put_env(@app, :api_mode, :watcher)`.

async means that two processes will execute the two tests concurrently, so tests not failing is luck :D

Comment (macaptain) [reply apps/omg_watcher_rpc/test/omg_watcher_rpc/tracer_test.exs:18]: Whoops, let me fix this.

File: apps/omg_watcher_rpc/lib/web/response.ex [hunk 5]
```diff
@@ -20,26 +20,21 @@ defmodule OMG.WatcherRPC.Web.Response do
...
-
-  defp service_name() do
-    case Application.get_env(@app, :api_mode) do
-      :watcher -> "watcher"
-      :watcher_info -> "watcher_info"
-    end
+    |> Map.put(:version, Configuration.version())
+    |> Map.put(
+      :service_name,
+      case Configuration.service_name() do
```
Comment (ayrat555) [inline apps/omg_watcher_rpc/lib/web/response.ex:34]: nitpick. you can use `to_string`

**Review by ayrat555 — APPROVED** [review]

File: apps/omg_watcher_rpc/lib/web/response.ex (hunk 5 above)
Comment (InoMurko) [reply apps/omg_watcher_rpc/lib/web/response.ex:34]: touche! or interpolation https://thepugautomatic.com/2016/01/elixir-string-interpolation-for-the-rubyist/
```
"#{Configuration.service_name()}"
```

**Pull request merged**